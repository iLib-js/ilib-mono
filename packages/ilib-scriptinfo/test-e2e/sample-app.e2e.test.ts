/*
 * sample-app.e2e.test.ts - E2E tests for ilib-scriptinfo sample app
 *
 * Copyright © 2025, Box, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import path from "path";
import { execFile } from "child_process";
import { promisify } from "util";

const execFileAsync = promisify(execFile);

interface TestResult {
    stdout: string;
    stderr: string;
    code: number;
}

/**
 * Sample app type
 */
type SampleType = "esm" | "legacy" | "typescript";

/**
 * Helper function to run the sample app with given arguments
 */
async function runSampleApp(args: string[] = [], sampleType: SampleType = "esm"): Promise<TestResult> {
    const sampleDir = path.resolve(__dirname, "..", "samples", sampleType);

    let command: string;
    let commandArgs: string[];

    if (sampleType === "typescript") {
        // For TypeScript, use tsx to run directly
        command = "tsx";
        commandArgs = [path.resolve(sampleDir, "index.ts"), ...args];
    } else {
        // For ESM and Legacy, use node with index.js
        const sampleAppPath = path.resolve(sampleDir, "index.js");
        command = "node";
        commandArgs = [sampleAppPath, ...args];
    }

    try {
        const { stdout, stderr } = await execFileAsync(command, commandArgs, {
            cwd: sampleDir,
            env: { ...process.env, NODE_ENV: "test" }
        });
        return { stdout, stderr, code: 0 };
    } catch (error: any) {
        return {
            stdout: error.stdout || "",
            stderr: error.stderr || "",
            code: error.code || 1
        };
    }
}

/**
 * Test suite for sample apps
 */
describe("ilib-scriptinfo sample apps", () => {
    // Test ESM, Legacy, and TypeScript samples
    const sampleTypes: SampleType[] = ["esm", "legacy", "typescript"];

    describe.each(sampleTypes)(`%s sample app`, (sampleType) => {

            describe("help functionality", () => {

                test("should show help when no arguments provided", async () => {
                    const result = await runSampleApp([], sampleType);
                    expect(result.code).toBe(0);
                    expect(result.stdout).toContain("ilib-scriptinfo");
                    expect(result.stdout).toContain("Sample App");
                    expect(result.stdout).toContain("USAGE");
                    expect(result.stdout).toContain("node index.js <script-code>");
                    expect(result.stdout).toContain("EXAMPLES");
                    expect(result.stdout).toContain("node index.js Latn");
                });

                test("should show help when --help flag is provided", async () => {
                    const result = await runSampleApp(["--help"], sampleType);
                    expect(result.code).toBe(0);
                    expect(result.stdout).toContain("ilib-scriptinfo");
                    expect(result.stdout).toContain("Sample App");
                    expect(result.stdout).toContain("USAGE");
                    expect(result.stdout).toContain("node index.js <script-code>");
                });

            });

            describe("valid script codes", () => {

                test("should display Latin script information correctly", async () => {
                    const result = await runSampleApp(["Latn"], sampleType);
                    expect(result.code).toBe(0);
                    expect(result.stdout).toContain('Script Information for "Latn"');
                    expect(result.stdout).toContain("Code             | Latn");
                    expect(result.stdout).toContain("Code Number      | 215");
                    expect(result.stdout).toContain("Name             | Latin");
                    expect(result.stdout).toContain("Script Direction | 📝 LTR Left-to-Right");
                    expect(result.stdout).toContain("IME Requirement  | ⌨️  No IME required");
                    expect(result.stdout).toContain("Casing Info      | 🔤 Uses letter case");
                });

                test("should display Arabic script information correctly", async () => {
                    const result = await runSampleApp(["Arab"], sampleType);
                    expect(result.code).toBe(0);
                    expect(result.stdout).toContain('Script Information for "Arab"');
                    expect(result.stdout).toContain("Code             | Arab");
                    expect(result.stdout).toContain("Name             | Arabic");
                    expect(result.stdout).toContain("Script Direction | 📝 RTL Right-to-Left");
                });

                test("should display Han script information correctly", async () => {
                    const result = await runSampleApp(["Hani"], sampleType);
                    expect(result.code).toBe(0);
                    expect(result.stdout).toContain('Script Information for "Hani"');
                    expect(result.stdout).toContain("Code             | Hani");
                    expect(result.stdout).toContain("Name             | Han");
                    expect(result.stdout).toContain("Script Direction | 📝 LTR Left-to-Right");
                });

            });

            describe("case correction functionality", () => {

                test("should correct lowercase input to proper case", async () => {
                    const result = await runSampleApp(["latn"], sampleType);
                    expect(result.code).toBe(0);
                    expect(result.stdout).toContain('Script Information for "latn"');
                    expect(result.stdout).toContain("Code             | Latn");
                    expect(result.stdout).toContain("Name             | Latin");
                });

                test("should correct uppercase input to proper case", async () => {
                    const result = await runSampleApp(["LATN"], sampleType);
                    expect(result.code).toBe(0);
                    expect(result.stdout).toContain('Script Information for "LATN"');
                    expect(result.stdout).toContain("Code             | Latn");
                    expect(result.stdout).toContain("Name             | Latin");
                });

            });

            describe("partial matching functionality", () => {

                test("should find partial matches for 'lat'", async () => {
                    const result = await runSampleApp(["lat"], sampleType);
                    expect(result.code).toBe(0);
                    expect(result.stdout).toContain("❌ Unknown script code: \"lat\"");
                    expect(result.stdout).toContain("🔍 Found");
                    expect(result.stdout).toContain("similar script code(s):");
                    expect(result.stdout).toContain("Latn - Latin");
                });

                test("should find partial matches for 'arab'", async () => {
                    const result = await runSampleApp(["arab"], sampleType);
                    // Both samples find exact match (case-insensitive)
                    expect(result.stdout).toContain('Script Information for "arab"');
                    expect(result.stdout).toContain("Code             | Arab");
                    expect(result.stdout).toContain("Name             | Arabic");
                });

                test("should find partial matches for 'han'", async () => {
                    const result = await runSampleApp(["han"], sampleType);
                    expect(result.code).toBe(0);
                    expect(result.stdout).toContain("❌ Unknown script code: \"han\"");
                    expect(result.stdout).toContain("🔍 Found");
                    expect(result.stdout).toContain("similar script code(s):");
                    expect(result.stdout).toContain("Hani - Han");
                });

            });

            describe("error handling", () => {

                test("should handle unknown script code with no matches", async () => {
                    const result = await runSampleApp(["xyzq"], sampleType);
                    expect(result.code).toBe(0);
                    expect(result.stdout).toContain("❌ Unknown script code: \"xyzq\"");
                    expect(result.stdout).toContain("💡 No similar script codes found");
                    expect(result.stdout).toContain("💡 Try one of these valid script codes:");
                    expect(result.stdout).toContain("Arab - Arabic");
                });

                test("should show error for too many arguments", async () => {
                    const result = await runSampleApp(["Latn", "Arab"], sampleType);
                    expect(result.code).toBe(1);
                    expect(result.stderr).toContain("Error: Invalid number of arguments");
                    expect(result.stderr).toContain("Usage: node index.js <script-code>");
                    expect(result.stderr).toContain("Example: node index.js Latn");
                });

            });

        });
    });

});