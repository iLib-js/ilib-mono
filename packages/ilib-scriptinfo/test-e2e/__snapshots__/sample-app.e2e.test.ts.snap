// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ilib-scriptinfo sample app case correction functionality should correct lowercase input to proper case 1`] = `
"
Script Information for "latn"
==================================================
Code             | Latn
Code Number      | 215
Name             | Latin
Long Code        | Latin
Script Direction | üìù LTR Left-to-Right
IME Requirement  | ‚å®Ô∏è  No IME required
Casing Info      | üî§ Uses letter case (uppercase/lowercase)
==================================================


"
`;

exports[`ilib-scriptinfo sample app case correction functionality should correct uppercase input to proper case 1`] = `
"
Script Information for "LATN"
==================================================
Code             | Latn
Code Number      | 215
Name             | Latin
Long Code        | Latin
Script Direction | üìù LTR Left-to-Right
IME Requirement  | ‚å®Ô∏è  No IME required
Casing Info      | üî§ Uses letter case (uppercase/lowercase)
==================================================


"
`;

exports[`ilib-scriptinfo sample app error handling should handle unknown script code with no matches 1`] = `
"
‚ùå Unknown script code: "xyz"
üí° No similar script codes found for "xyz"
üí° Try one of these valid script codes:
   Adlm - Adlam
   Afak - Afaka
   Aghb - Caucasian Albanian
   Ahom - Ahom, Tai Ahom
   Arab - Arabic
   Aran - Arabic (Nastaliq variant)
   Armi - Imperial Aramaic
   Armn - Armenian
   Avst - Avestan
   Bali - Balinese
   ... and 216 more scripts


"
`;

exports[`ilib-scriptinfo sample app error handling should show error for too many arguments 1`] = `
"Error: Invalid number of arguments

Usage: node index.js <script-code>
Example: node index.js Latn
Example: node index.js Arab
Example: node index.js Hani

For more information, run: node index.js --help
"
`;

exports[`ilib-scriptinfo sample app help functionality should show help when --help flag is provided 1`] = `
"
ilib-scriptinfo Sample App
==========================

A command-line tool that demonstrates the ilib-scriptinfo package by looking up
information about writing scripts based on ISO 15924 codes.

USAGE
-----
node index.js <script-code>
node index.js --help

PARAMETERS
----------
<script-code>    ISO 15924 4-letter script code (e.g., Latn, Arab, Hani)

OPTIONS
-------
--help           Show this help message

EXAMPLES
--------
node index.js Latn          # Look up Latin script information
node index.js Arab          # Look up Arabic script information
node index.js Hani          # Look up Han (Chinese) script information
node index.js Deva          # Look up Devanagari script information
node index.js Cyrl          # Look up Cyrillic script information

OUTPUT
------
The tool displays all available script properties in a tabular format:
‚Ä¢ Code: ISO 15924 4-letter script code
‚Ä¢ Code Number: ISO 15924 numeric code
‚Ä¢ Name: English name of the script
‚Ä¢ Long Code: Long identifier for the script
‚Ä¢ Script Direction: Writing direction (ltr/rtl)
‚Ä¢ Needs IME: Whether script requires Input Method Editor
‚Ä¢ Uses Casing: Whether script uses letter case

COMMON SCRIPT CODES
------------------
Latn - Latin
Arab - Arabic
Hani - Han (Chinese)
Deva - Devanagari
Cyrl - Cyrillic
Grek - Greek
Hebr - Hebrew
Thai - Thai
Hang - Hangul (Korean)
Kana - Katakana
Hira - Hiragana

For a complete list of all available script codes, the tool will show
examples when an unknown script code is provided.

MORE INFORMATION
----------------
This is a sample application demonstrating the ilib-scriptinfo package.
For more information about the package, see the main documentation.

"
`;

exports[`ilib-scriptinfo sample app help functionality should show help when no arguments provided 1`] = `
"
ilib-scriptinfo Sample App
==========================

A command-line tool that demonstrates the ilib-scriptinfo package by looking up
information about writing scripts based on ISO 15924 codes.

USAGE
-----
node index.js <script-code>
node index.js --help

PARAMETERS
----------
<script-code>    ISO 15924 4-letter script code (e.g., Latn, Arab, Hani)

OPTIONS
-------
--help           Show this help message

EXAMPLES
--------
node index.js Latn          # Look up Latin script information
node index.js Arab          # Look up Arabic script information
node index.js Hani          # Look up Han (Chinese) script information
node index.js Deva          # Look up Devanagari script information
node index.js Cyrl          # Look up Cyrillic script information

OUTPUT
------
The tool displays all available script properties in a tabular format:
‚Ä¢ Code: ISO 15924 4-letter script code
‚Ä¢ Code Number: ISO 15924 numeric code
‚Ä¢ Name: English name of the script
‚Ä¢ Long Code: Long identifier for the script
‚Ä¢ Script Direction: Writing direction (ltr/rtl)
‚Ä¢ Needs IME: Whether script requires Input Method Editor
‚Ä¢ Uses Casing: Whether script uses letter case

COMMON SCRIPT CODES
------------------
Latn - Latin
Arab - Arabic
Hani - Han (Chinese)
Deva - Devanagari
Cyrl - Cyrillic
Grek - Greek
Hebr - Hebrew
Thai - Thai
Hang - Hangul (Korean)
Kana - Katakana
Hira - Hiragana

For a complete list of all available script codes, the tool will show
examples when an unknown script code is provided.

MORE INFORMATION
----------------
This is a sample application demonstrating the ilib-scriptinfo package.
For more information about the package, see the main documentation.

"
`;

exports[`ilib-scriptinfo sample app partial matching functionality should find partial matches for 'arab' 1`] = `
"
Script Information for "arab"
==================================================
Code             | Arab
Code Number      | 160
Name             | Arabic
Long Code        | Arabic
Script Direction | üìù RTL Right-to-Left
IME Requirement  | ‚å®Ô∏è  No IME required
Casing Info      | üî§ No letter case
==================================================


"
`;

exports[`ilib-scriptinfo sample app partial matching functionality should find partial matches for 'han' 1`] = `
"
‚ùå Unknown script code: "han"
üîç Found 12 similar script code(s):
   Dupl - Duployan shorthand, Duployan stenography
   Hanb - Han with Bopomofo (alias for Han + Bopomofo)
   Hang - Hangul (Hang≈≠l, Hangeul)
   Hani - Han (Hanzi, Kanji, Hanja)
   Hano - Hanunoo (Hanun√≥o)
   Hans - Han (Simplified variant)
   Hant - Han (Traditional variant)
   Hntl - Han (Traditional variant) with Latin (alias for Hant + Latn)
   Jamo - Jamo (alias for Jamo subset of Hangul)
   Jpan - Japanese (alias for Han + Hiragana + Katakana)
   Kore - Korean (alias for Hangul + Han)
   Rohg - Hanifi Rohingya


"
`;

exports[`ilib-scriptinfo sample app partial matching functionality should find partial matches for 'lat' 1`] = `
"
‚ùå Unknown script code: "lat"
üîç Found 4 similar script code(s):
   Hntl - Han (Traditional variant) with Latin (alias for Hant + Latn)
   Latf - Latin (Fraktur variant)
   Latg - Latin (Gaelic variant)
   Latn - Latin


"
`;

exports[`ilib-scriptinfo sample app valid script codes should display Arabic script information correctly 1`] = `
"
Script Information for "Arab"
==================================================
Code             | Arab
Code Number      | 160
Name             | Arabic
Long Code        | Arabic
Script Direction | üìù RTL Right-to-Left
IME Requirement  | ‚å®Ô∏è  No IME required
Casing Info      | üî§ No letter case
==================================================


"
`;

exports[`ilib-scriptinfo sample app valid script codes should display Han script information correctly 1`] = `
"
Script Information for "Hani"
==================================================
Code             | Hani
Code Number      | 500
Name             | Han (Hanzi, Kanji, Hanja)
Long Code        | Han
Script Direction | üìù LTR Left-to-Right
IME Requirement  | ‚å®Ô∏è  Requires Input Method Editor
Casing Info      | üî§ No letter case
==================================================


"
`;

exports[`ilib-scriptinfo sample app valid script codes should display Latin script information correctly 1`] = `
"
Script Information for "Latn"
==================================================
Code             | Latn
Code Number      | 215
Name             | Latin
Long Code        | Latin
Script Direction | üìù LTR Left-to-Right
IME Requirement  | ‚å®Ô∏è  No IME required
Casing Info      | üî§ Uses letter case (uppercase/lowercase)
==================================================


"
`;
